javascript/newer-JS-features.md

```diff
@@ -0,0 +1,219 @@
+const secondaryColors = ['purple','pink', 'crimson'];
+const allColors = [...primaryColors, ...secondaryColors] // (6)["red", "green", "blue", "purple", "pink", "crimson"]
+```
+### Spread with Objects
+Spread in object literal copies properties from one object into another object. In past, we've used Object.assign to copy properties from one object to another. Spread makes copying more easy!
+```javascript
+const personA = {
+    favColor:"pink"
+}
+const personB = {
+    favFood:"pasta"
+}
+const info = {...personA, ...personB}
+console.log(info); // {favColor: "pink", favFood: "pasta"}
+```
+In case of conflict where both objects have similar properties, order matters! 
+```javascript
+const personC = {
+    favColor:"pink",
+    favPet:"cat"
+}
+const personD = {
+    favFood:"pasta",
+    favPet:"dog"
+}
+const anotherInfo = {...personA, ...personB } // {favColor: "pink", favPet: "dog", favFood: "pasta"}
+```
+## Rest parameters...
+Rest parameters allows a function to accept an indefinite numbers of arguments as an array. Rest looks like spread, but it’s NOT spread. Before taking a look at Rest, let's take a look at **Arguments Object** which  is an array like object but it doesn't have access to array like methods like map, reduce. Also, it's not available with arrow function. It automatically hold all the values passed to the function.
+## Examples
+```javascript
+function sum() {
+    console.log(arguments) // collects all arguments passed to the function. 
+}
+sum(2,3,4,5) // Arguments(3) [2, 3, 4, callee: ƒ, Symbol(Symbol.iterator): ƒ]
+```
+It contains all values that we passed into the function by order, but we can't use array like methods like reduce. Here, let's write a function that will give sum of all the elements in the array.
+```javascript
+function sum() {
+    return arguments.reduce((total, acc) => total + acc);
+} 
+sum(2,3,4,5,6) // arguments.reduce is not a function
+```
+This is where Rest comes in. Rest  collects all remaining arguments or rest of the arguments and puts them into an array. The dots means gather the rest of the parameters into an array, so the rest parameters must be at the end. Rest is the opposite of Spread. In spread we're spreading things out, but in Rest we're collecting all things into a single parameter. 
+
+```javascript
+function sumOfAllArray (...arr) {
+    let total = 0;
+    for (let i of arr) total +=i;
+    return total;
+}
+console.log("The sum of array is: ", sumOfAllArray(3,4,5));  // 12
+```
+
+## Array Destructuring Assignment
+Destructuring assignment syntax allows us to unpack or extract or single out values from arrays, properties from objects into distinct variables. With array destructuring, we can copy items into variables without modifying the original array.
+
+## Examples
+### Array Destructuring
+Let's look at numbers array. variable max contains the first element of the array and secondMax contains the second element of the array and so on. Old way of doing it would be: 
```

Comment:
```suggestion
Let's look at numbers array. Variable `max` contains the first element of the array and `secondMax` contains the second element of the array and so on. Old way of doing it would be: 
```